lib/faraday/cookie_jar.rb

```diff
@@ -6,25 +6,30 @@ class CookieJar < Faraday::Middleware
     def initialize(app, options = {})
       super(app)
       @jar = options[:jar] || HTTP::CookieJar.new
+      @mutex = Mutex.new
     end
 
     def call(env)
-      cookies = @jar.cookies(env[:url])
-      unless cookies.empty?
-        cookie_value = HTTP::Cookie.cookie_value(cookies)
-        if env[:request_headers]["Cookie"]
-          unless env[:request_headers]["Cookie"] == cookie_value
-            env[:request_headers]["Cookie"] = cookie_value + ';' + env[:request_headers]["Cookie"]
+      @mutex.synchronize do
+        cookies = @jar.cookies(env[:url])
+        unless cookies.empty?
```

Comment:
do you still need mutex after retrieving the cookie? The following lines do not touch the storage of the cookies.

Reply:
I couldn't quite tell whether http-cookie guarantees that it always returns a new object isntead of a reference to a list stored inside itself, so I decided to play it safe. Since the "extra" lines are just manipulating a dictionary, I figured that the extra few nanoseconds inside the mutex wouldn't be a big deal.

I'm open to removing it if you are confident that `http-cookie` never returns references